src/server/frontend_xwayland/xwayland_wm_surface.h

```diff
@@ -122,38 +123,53 @@ class XWaylandWMShellSurface;
 class XWaylandWMSurface
 {
 public:
-    enum WmState
-    {
-        WithdrawnState = 0,
-        NormalState = 1,
-        IconicState = 3
-    };
-
     XWaylandWMSurface(XWaylandWM *wm, xcb_create_notify_event_t *event);
     ~XWaylandWMSurface();
+    void net_wm_state_client_message(uint32_t const (&data)[5]);
+    void wm_change_state_client_message(uint32_t const (&data)[5]);
     void dirty_properties();
     void read_properties();
     void set_surface(WlSurface* wayland_surface); ///< Should only be called on the Wayland thread
     void set_workspace(int workspace);
-    void set_wm_state(WmState state);
-    void set_net_wm_state();
+    void apply_mir_state_to_window(MirWindowState new_state);
+    void unmap();
     void move_resize(uint32_t detail);
-    void set_state(MirWindowState state);
     void send_resize(const geometry::Size& new_size);
     void send_close_request();
 
 private:
     /// Runs work on the Wayland thread if the shell surface hasn't been destroyed
     void aquire_shell_surface(std::function<void(XWaylandWMShellSurface* shell_surface)>&& work);
 
+    /// contains more information than just a MirWindowState
+    /// (for example if a minimized window would otherwise be maximized)
+    struct WindowState
```

Comment:
Why does this exist? This information *should* be stored in Mir's internal representation, because this is a core WM concern, not something specific to X11.

Reply:
> This information _should_ be stored in Mir's internal representation, because this is a core WM concern, not something specific to X11.

Agreed. But maybe, pending more extensive changes to the "model", storing it in the "view" works well enough?

(We've been holding off reworking the "model" until we clean out mirclient dependencies.)

Reply:
> This information should be stored in Mir's internal representation

I would need to be convinced of this. See longer comment below.